packages/@aws-cdk/aws-ecs/README.md

```diff
@@ -589,3 +589,59 @@ taskDefinition.addContainer('TheContainer', {
   })
 });
 ```
+
+## Capacity Provider
+
+### Autoscaling Group Capacity Provider
+
+Amazon ECS capacity providers can use Auto Scaling groups to manage the Amazon EC2 instances registered to
+their clusters. You can use the managed scaling feature to have Amazon ECS manage the scale-in and scale-out
+actions of the Auto Scaling group or you can manage the scaling actions yourself. This enables you to effectively
+use cluster auto scaling.
+
+The following sample creates two Autoscaling Group Capacity Providers and register them to the cluster.
+
+```ts
+const cluster = new ecs.Cluster(stack, 'Cluster', { vpc });
+
+// create the 1st capacity provider with on-demand t3.large instances
+const cp = cluster.addCapacityProvider('CP', {
```

Comment:
A more general question: is it possible to merge `addCapacityProviderConfiguration` into `addCapacityProvider` so that users don't need to create a capacity provider then register it to a cluster? Because I would expect `cluster.addCapacityProvider` already create and register a capacity provider to that cluster and return the created capacity provider (which is similar to what users would expect for `cluster.addCapacity`)

Reply:
> A more general question: is it possible to merge `addCapacityProviderConfiguration` into `addCapacityProvider` so that users don't need to create a capacity provider then register it to a cluster? Because I would expect `cluster.addCapacityProvider` already create and register a capacity provider to that cluster and return the created capacity provider (which is similar to what users would expect for `cluster.addCapacity`)

Hi @iamhopaul123 

I have been trying to explore this but didn't make it yet. Let me explain my current status and I am looking for your advice.
`
`addCapacityProvider()` is easy as we just need to create the `AWS::ECS::CapacityProvider` resource, but register multiple capacity providers to the cluster is still challenging. We probably have two options here:

1. simply associate the cp to the cluster in the `CapacityProvider`  and `DefaultCapacityProviderStrategy` properties of the `AWS::ECS::Cluster`
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecs-cluster.html#cfn-ecs-cluster-capacityproviders

However, we'll have circular dependency(https://github.com/aws/containers-roadmap/issues/631#issuecomment-663153933). An option is like this https://github.com/aws/containers-roadmap/issues/631#issuecomment-671129605 but it might cause breaking changes.

2. The second options is what I've been working on now. I am creating a new `CapacityProviderConfiguration` custom resource. When we `addCapacityProvider()`, the cp will be added into `capacityProviders: CapacityProviders[]` of the cluster instance and on cluster creation, we synthesize all the added capacity providers  and run the `new CapacityProviderConfiguration` inside the `Cluster` class, but in this case the `CapacityProviderConfiguration` custom resource can hardly addDependency on all the capacity providers, which will lead to the error below:

>Failed to delete resource. Error: An error occurred (UpdateInProgressException) when calling the PutClusterCapac
ityProviders operation: The specified cluster is in a busy state. Cluster attachments must be in UPDATE_COMPLETE
or UPDATE_FAILED state before they can be updated. Wait and try again.

I believe the 2nd option is the way to go, but I am still struggling with how to addDependency on the capacity providers for the `CapacityProviderConfiguration` resource.

I am not sure if I am on the right track and would be appreciated if you can share your feedbacks. What do you think?

Reply:
RE the circular dependency.

I believe we need break the following 2 circular dependencies:

cluster -> cp -> addCapacity() -> addAutoScalingGroup() -> cluster

https://github.com/aws/aws-cdk/blob/d31d4db555d1a87e45d6ddd4c3bbea84b4854e3a/packages/%40aws-cdk/aws-ecs/lib/cluster.ts#L198-L199

cluster -> cp -> addCapacity() -> addAutoScalingGroup() -> InstanceDrainHook -> cluster

https://github.com/aws/aws-cdk/blob/d31d4db555d1a87e45d6ddd4c3bbea84b4854e3a/packages/%40aws-cdk/aws-ecs/lib/cluster.ts#L258-L265

Reply:
OK. Now I am able to resolve the two circular dependencies.

Will add comments inline for discussion.